queue/queue.go

```diff
@@ -16,47 +18,431 @@ var DefaultBufferSize = runtime.GOMAXPROCS(0) * 10
+// all pending messages.
+func (q *Queue) pushPending(new bool, e *elem) {
+	q.bufferM.Lock()
+	defer q.bufferM.Unlock()
+
+	oversize := false
+
+	// If the message is "new" it's not in the buffer yet.
+	if new {
+		if q.buffered == nil {
+			q.buffered = map[string]struct{}{}
+		}
+
+		q.buffered[e.message.ID()] = struct{}{}
+
+		size := q.BufferSize
+		if size <= 0 {
+			size = DefaultBufferSize
+		}
+
+		if len(q.buffered) > size {
+			// The buffer has exceeded its size limit, we need to discard the
+			// lowest priority element.
+			oversize = true
+		}
+	}
+
+	if q.pending.Push(e) {
+		q.wakeMonitor()
+	}
+
+	if oversize {
+		// We've got more messages in memory than the limit allows, drop the
+		// lowest-priority element.
+		//
+		// Note: There's a chance this is the message we just pushed. That would
+		// mean that all buffered messages are in active sessions, which should
+		// not occur if BufferSize is larger than the number of consumers as
+		// suggested by its documentation.
+		drop, _ := q.pending.PopBack()
+		delete(q.buffered, drop.(*elem).message.ID())
+
+		// There are now persisted messages that are not in the buffer.
+		q.exhaustive = false
+	}
+}
+
+// loadMessages reads queued messages from the data-store into the buffer.
+//
+// If messages are loaded, and the highest-priority message is ready to be
+// handled now, e is an element containing that message and ok is true.
+func (q *Queue) loadMessages(ctx context.Context) (e *elem, ok bool, err error) {
+	size := q.BufferSize
+	if size <= 0 {
+		size = DefaultBufferSize
+	}
+
+	// Load just enough messages to fill the buffer.
+	messages, err := q.DataStore.QueueStoreRepository().LoadQueueMessages(ctx, size)
+
+	q.bufferM.Lock()
+	defer q.bufferM.Unlock()
+
+	// Clear the loading flag regardless of whether there was an error.
+	q.loading = false
+
 	if err != nil {
-		return err
+		return nil, false, err
 	}
-	defer tx.Rollback()
 
-	if err := tx.SaveMessageToQueue(ctx, it.message); err != nil {
-		return err
+	if len(messages) < size {
+		// We didn't get enough messages to fill the buffer, so we know we have
+		// everything in memory.
+		q.exhaustive = true
 	}
 
-	return tx.Commit(ctx)
+	// In fact, we got nothing at all!
+	if len(messages) == 0 {
+		return nil, false, nil
+	}
+
+	// If the message at the front of the queue is ready now we are going to
+	// return it to our caller directly to avoid starting the monitor goroutine.
+	//
+	// We remove it from the slice so that it's not added to the pending list.
+	m := messages[0]
+	if !m.NextAttemptAt.After(time.Now()) {
+		e = &elem{message: m}
+		messages = messages[1:]
+	}
+
+	// Add an element to the buffer for each message that we won't be returning
+	// to the caller, then wake the monitoring goroutine.
+	if len(messages) > 0 {
+		for _, m := range messages {
+			if q.buffered == nil {
+				q.buffered = map[string]struct{}{}
+			}
+			q.buffered[m.ID()] = struct{}{}
+			q.pending.Push(&elem{message: m})
```

Comment:
Need to check for duplicates here, as a goroutine may have called `Push()` while we were loading.

Reply:
Tested and hopefully fixed in 8223190.